arch/arm/core/aarch32/cpu_idle.S

```diff
@@ -48,17 +48,40 @@ SECTION_FUNC(TEXT, z_arm_cpu_idle_init)
 #endif
 	bx	lr
 
-SECTION_FUNC(TEXT, arch_cpu_idle)
-#ifdef CONFIG_TRACING
+.macro _prologue
+#if defined(CONFIG_TRACING) || defined(CONFIG_ARM_ON_ENTER_CPU_IDLE_HOOK)
 	push	{r0, lr}
+#if defined(CONFIG_TRACING)
 	bl	sys_trace_idle
+#endif
+#if defined(CONFIG_ARM_ON_ENTER_CPU_IDLE_HOOK)
+	bl	z_arm_on_enter_cpu_idle
+	/* Preserve the return value so that it can be used later,
+	 * see the _sleep_if_allowed macro.
+	 */
+	mov	r2, r0
+#endif
 #if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE)
 	pop	{r0, r1}
 	mov	lr, r1
 #else
 	pop	{r0, lr}
 #endif /* CONFIG_ARMV6_M_ARMV8_M_BASELINE */
-#endif /* CONFIG_TRACING */
+#endif /* CONFIG_TRACING || CONFIG_ARM_ON_ENTER_CPU_IDLE_HOOK */
+.endm
+
+.macro _sleep_if_allowed wait_instruction
+#if defined(CONFIG_ARM_ON_ENTER_CPU_IDLE_HOOK)
+	/* Skip the wait instruction if on_enter_cpu_idle() returned false. */
+	cmp	r2, #0
+	beq	_skip_instruction_\@
+#endif
+	\wait_instruction
+_skip_instruction_\@:
+.endm
```

Comment:
Same here. We really do not need to make two instructions a macro ...

Reply:
Agree, it does feel a bit overkill

Reply:
I reworked this a bit. Now the `_sleep_if_allowed` macro contains more instructions. This sequence of code is used three times in the file, so I think using a macro in this case makes sense.